collectors/ebpf.plugin/ebpf_functions.c

```diff
@@ -0,0 +1,413 @@
+            buffer_json_member_add_array(wb, "columns");
+            {
+                buffer_json_add_array_item_string(wb, "Status");
+            }
+            buffer_json_array_close(wb);
+        }
+        buffer_json_object_close(wb);
+    }
+    buffer_json_object_close(wb); // group_by
+
+    buffer_json_member_add_time_t(wb, "expires", expires);
+    buffer_json_finalize(wb);
+
+    // Lock necessary to avoid race condition
+    pthread_mutex_lock(&lock);
+    fwrite(buffer_tostring(wb), buffer_strlen(wb), 1, stdout);
+
+    pluginsd_function_result_end_to_stdout();
+    fflush(stdout);
+    pthread_mutex_unlock(&lock);
+
+    buffer_free(wb);
+}
+
+
+/*****************************************************************
+ *  EBPF FUNCTION THREAD
+ *****************************************************************/
+
+/**
+ * FUNCTION thread.
+ *
+ * @param ptr a `ebpf_module_t *`.
+ *
+ * @return always NULL.
+ */
+void *ebpf_function_thread(void *ptr)
+{
+    ebpf_module_t *em = (ebpf_module_t *)ptr;
+    char buffer[PLUGINSD_LINE_MAX + 1];
+
+    char *s = NULL;
+    while(!ebpf_exit_plugin && (s = fgets(buffer, PLUGINSD_LINE_MAX, stdin))) {
+        char *words[PLUGINSD_MAX_WORDS] = { NULL };
+        size_t num_words = quoted_strings_splitter_pluginsd(buffer, words, PLUGINSD_MAX_WORDS);
+
+        const char *keyword = get_word(words, num_words, 0);
+
+        if(keyword && strcmp(keyword, PLUGINSD_KEYWORD_FUNCTION) == 0) {
+            char *transaction = get_word(words, num_words, 1);
+            char *timeout_s = get_word(words, num_words, 2);
+            char *function = get_word(words, num_words, 3);
+
+            if(!transaction || !*transaction || !timeout_s || !*timeout_s || !function || !*function) {
+                netdata_log_error("Received incomplete %s (transaction = '%s', timeout = '%s', function = '%s'). Ignoring it.",
+                                  keyword,
+                                  transaction?transaction:"(unset)",
+                                  timeout_s?timeout_s:"(unset)",
+                                  function?function:"(unset)");
+            }
+            else {
+                int timeout = str2i(timeout_s);
+                if (!strncmp(function, EBPF_FUNCTION_THREAD, sizeof(EBPF_FUNCTION_THREAD) - 1))
+                    ebpf_function_thread_manipulation(transaction,
+                                                      function,
+                                                      buffer,
+                                                      PLUGINSD_LINE_MAX + 1,
+                                                      timeout,
+                                                      em);
+                else
+                    ebpf_function_error(transaction,
+                                        HTTP_RESP_NOT_FOUND,
+                                        "No function with this name found in ebpf.plugin.");
```

Comment:
Hi Thiago!

Seems adding a `fflush(stdio);` here, seems to solve my issues with delay of getting `http://localhost:19999/api/v1/function?function=ebpf_thread%20help` (6ms vs 3-4 secs). Could you try as well?

Reply:
Thank you @MrZammler , I pushed a commit adding not only the `fflush`, but also the necessary `mutex` around.